src/main/java/org/opensearch/knn/index/codec/util/KNNVectorAsCollectionOfHalfFloatsSerializer.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.codec.util;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.knn.jni.SIMDDecoding;
+import org.opensearch.knn.jni.SIMDEncoding;
+
+import java.nio.ByteOrder;
+
+/**
+ * Class implements KNNVectorSerializer based on serialization/deserialization of float array
+ * as a collection of individual half-precision values
+ */
+public class KNNVectorAsCollectionOfHalfFloatsSerializer {
+    private static final Logger log = LogManager.getLogger(KNNVectorAsCollectionOfHalfFloatsSerializer.class);
+
+    private static final int BYTES_IN_HALF_FLOAT = 2;
+    private static final boolean IS_LITTLE_ENDIAN = ByteOrder.nativeOrder().equals(ByteOrder.LITTLE_ENDIAN);
+
+    public static final KNNVectorAsCollectionOfHalfFloatsSerializer INSTANCE = new KNNVectorAsCollectionOfHalfFloatsSerializer();
+
+    /**
+     * Converts float[] to byte[] using SIMD optimization if supported, otherwise fallback to Java.
+     * @param input the float[] to be serialized into half-precision format
+     * @param output byte[] containing the float16-encoded data
+     * @param dimension number of floats to serialize
+     */
+    public void floatToByteArray(float[] input, byte[] output, int dimension) {
+        if (input == null || output == null) {
+            throw new IllegalArgumentException("Input/output buffers cannot be null.");
+        }
+        if (dimension > input.length) {
+            throw new IllegalArgumentException("Count exceeds input float array length.");
+        }
+        if (output.length != input.length * BYTES_IN_HALF_FLOAT) {
+            throw new IllegalArgumentException("Output buffer size mismatch. Must be 2x input length.");
+        }
+
+        if (SIMDEncoding.isSIMDSupported() && SIMDEncoding.convertFP32ToFP16(input, output, dimension)) {
```

Comment:
I think we'd also benefit from implementing a fallback approach here.
Currently we throw out the exception in case of CPP failures and could lead to cascading failures.

We can work it similar to what we have for remote index builds - 

```
if (SIMDEncoding.isSIMDSupported()) {
    try {
        SIMDEncoding.convertFP32ToFP16(input, output, dimension);
        return; 
    } catch (Exception e) {
        logger.debug("SIMD conversion failed, falling back to Java: {}", e.getMessage());
        // Continue to Java fallback
    }
}
```